doc/use_cli_effectively.md

```diff
@@ -0,0 +1,88 @@
+# Tips of using Azure CLI effectively #
+
+## Use the right output mode for your work (json, table, or tsv) ##
+  1. Use `json` (CLI's default) to give you the most comprehensive information. If you prefer less, use `--output` argument to override at individual command level, or use "az configure" for global level.
+  2. Use `table` to give you less but much focused information. If some fields you care are missing, you have 2 options, use `--output json` to get all, or leverage `--query` to pick ones you like.
+  
+         az vm show -g my_rg -n my_vm --query "{name: name, os:storageProfile.imageReference.offer}" -otable
+         Name    Os
+         ------  ------------
+         my_vm   UbuntuServer
+         
+  3. Use `tsv` for the most concise output, which is good for scripting particularly you can save to variables to be used later, e.g.
+        
+        export vm_ids=$(az vm list -d -g my_rg --query "[?powerState=='VM running'].id" -o tsv)  
+        az vm stop --ids $vm_ids
+        
+## To pass values from one command to the other ##
+  1. If the value will be used more than once, assign it to a variable, e.g.
+       
+         running_vm_ids=$(az vm list -d -g my_rg --query "[?powerState=='VM running'].id" -o tsv)
+       
+     Or, if it gets used only once, you can just pipe
+
+         az vm list --query "[?powerState=='VM running'].name" | grep my_vm
+
+  2. For list type, 2 suggestions:
+
+       If you need more controls on the result, use "for" loop:
+       
+            #!/usr/bin/env bash
+            for vm in $(az vm list -d -g my_rg --query "[?powerState=='VM running'].id" -o tsv); do
+                echo stopping $vm
+                az vm stop --ids $vm
+                if [ $? -ne 0 ]; then
+                    echo "Failed to stop $vm"
+                    exit 1
+                fi
+                echo $vm stopped
+            done
+        
+       Or, use `xargs` and optionally going parallel through `-P` flag for better performance
+
+           az vm list -d -g my_rg --query "[?powerState=='VM stopped'].id" -o tsv | xargs -I {} -P 10 az vm start --ids "{}"
+
+       Less known, but CLI also has built-in support on processing "--ids" in parallel to achieve the same effect of xargs
+
+          az vm list -d -g my_rg --query "[?powerState=='VM stopped'].id" -o tsv | tr '\r\n' ' ' | { read ids ; az vm start --ids $ids ; }
+
+## Async executions ##
+  This becomes handy for a few scenarios:
+  1. Clean up
+
+          az group delete -n my_rg --no-wait
+
+  2. For performance, don't wait on each operation, rather wait for all at the end
+      
+          az vm create -g my_rg -n vm1 --image centos --no-wait
+          az vm create -g my_rg -n vm2 --image centos --no-wait
+           
+          az vm wait --created --ids $(az vm list -g my_rg --query "[].id" -o tsv)
```

Comment:
If this is executed in a script, it is possible that the new VMs might not appear in the list.